### PR TITLE
Feature/2062/facet search

### PR DESCRIPTION
**Pull request by nrzhao4**
- For [2062](https://github.com/dockstore/dockstore/issues/2062l)
- Originally made [this PR](https://github.com/dockstore/dockstore-ui2/pull/1085) but changed to a completely different approach
- filtering is done via pipes, rather than making actual changes to the entryOrder map

### CODE REVIEW

File: src/app/search/state/search.service.ts [hunk 1]
```diff
@@ -285,6 +285,21 @@ export class SearchService {
     });
   }
 
+  setFacetAutocompleteTerms(hits: any) {
+    const autocompleteTerms = [];
+    for (let i = 0; i <= 3; i++) {
```
Comment (coverbeck) [inline src/app/search/state/search.service.ts:290]: Where does 3 come from? Suggest a const with a meaningful name

File: src/app/search/search.component.ts [hunk 2]
```diff
@@ -516,6 +526,16 @@ export class SearchComponent implements OnInit, OnDestroy {
     this.sortModeMap.get(category).SortBy = sortMode;
   }
 
+  // Get autocomplete terms
+  onFacetSearchKey(key) {
+    const values = this.facetSearchText.toLowerCase();
+    const unfilteredItems = Array.from(this.orderedBuckets.get(key).Items.entries());
+    const filteredItems = unfilteredItems.filter((item) => item[0].toLowerCase().includes(values)).map((item) => item[0]);
+    if (filteredItems) {
```
Comment (coverbeck) [inline src/app/search/search.component.ts:534]: This will always be true. If you want the logic this seems to imply, you probably need to test that it's empty.

That said, what happens if it is empty? Do you not want to clear the existing terms (I don't know)?

File: src/app/search/state/search.service.ts (hunk 1 above)
Comment (nrzhao4) [reply src/app/search/state/search.service.ts:290]: Following the other autocomplete implementation where at most there are 4 terms

File: src/app/search/search.component.ts (hunk 2 above)
Comment (nrzhao4) [reply src/app/search/search.component.ts:534]: Oops, my logic went awry here. If its empty, the function setFacetAutocompleteTerms can still be called but will leave autocomplete array empty

File: src/app/search/facet-search/facet-search-update.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ *
+ * @param {Array<any>} items The items in the facet
+ * @param {string} searchText The search text entered
+ * @returns {string} The numbers of hidden items in the facet
+ * @memberof GetFacetSearchResultsPipe
+ */
+export class GetFacetSearchUpdatePipe implements PipeTransform {
+  transform(unfilteredItems: Array<any>, searchText: string): string {
+    let hiddenFilteredSize;
+    if (!searchText && unfilteredItems.length > 5) {
```
Comment (coverbeck) [inline src/app/search/facet-search/facet-search-update.pipe.ts:17]: Magic number 5 used 3 times in this method; should extract to a const. I thought our linter caught stuff like this.

File: src/app/search/facet-search/facet-search-update.pipe.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      hiddenFilteredSize = unfilteredItems.length - 5;
+      return hiddenFilteredSize.toString() + ' more';
+    }
+    const filteredSize = unfilteredItems.filter((item) => item.toLowerCase().includes(searchText)).length;
```
Comment (coverbeck) [inline src/app/search/facet-search/facet-search-update.pipe.ts:21]: Shouldn't you make a lower-case copy of searchText at the beginning of this method? Otherwise you're relying on the caller always passing in lower-case.

File: src/app/search/facet-search/facet-search-update.pipe.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+/**
+ * This pipe updates the number of hidden items in the facet when a search is completed
+ *
+ * @param {Array<any>} items The items in the facet
+ * @param {string} searchText The search text entered
+ * @returns {string} The numbers of hidden items in the facet
+ * @memberof GetFacetSearchResultsPipe
+ */
+export class GetFacetSearchUpdatePipe implements PipeTransform {
+  transform(unfilteredItems: Array<any>, searchText: string): string {
```
Comment (coverbeck) [inline src/app/search/facet-search/facet-search-update.pipe.ts:15]: Return type should be `string | null`, unless the interface doesn't allow it.

File: src/app/search/facet-search/facet-search.pipe.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ * @param {string} facet The facet the search applies to
+ * @returns {Array<any>} The results that match the search
+ * @memberof GetFacetSearchResultsPipe
+ */
+export class GetFacetSearchResultsPipe implements PipeTransform {
+  transform(items: Array<any>, searchText: string, facet: string): Array<any> {
+    if (!items || !searchText || facet !== 'author') {
+      return items;
+    }
+    const filteredItems = items.filter((item) => item.toLowerCase().indexOf(searchText.toLowerCase()) !== -1);
```
Comment (coverbeck) [inline src/app/search/facet-search/facet-search.pipe.ts:20]: 1. Micro-optimization that maybe the compiler already handles, but you would ideally create a variable out of `searchText.toLowerCase()` -- otherwise you are executing it again and again for every item in the array.
2. You could use `includes` instead of `indexOf` like you do in facet-search-update-pipe

**Review by coverbeck — COMMENTED** [review]
I ran this locally and it looks good to me!

Some nit-pick comments..

File: src/app/search/facet-search/facet-search-update.pipe.ts (hunk 3 above)
Comment (garyluu) [reply src/app/search/facet-search/facet-search-update.pipe.ts:17]: Only our Java checkstyle (not tests) seems to complain about magic numbers.  Javascript/Typescript linters/projects typically don't have that as a default unless otherwise specified

Comment (denis-yuen) [reply src/app/search/facet-search/facet-search-update.pipe.ts:17]: Can confirm I configured it for Java non-tests. 
Would be nice for the UI (non-tests) as well

File: src/app/search/facet-search/facet-search-update.pipe.ts [hunk 7]
```diff
@@ -12,17 +12,19 @@ import { Pipe, PipeTransform } from '@angular/core';
  * @memberof GetFacetSearchResultsPipe
  */
 export class GetFacetSearchUpdatePipe implements PipeTransform {
-  transform(unfilteredItems: Array<any>, searchText: string): string {
+  transform(unfilteredItems: Array<any>, searchText: string): string | null {
     let hiddenFilteredSize;
-    if (!searchText && unfilteredItems.length > 5) {
-      hiddenFilteredSize = unfilteredItems.length - 5;
+    const value = searchText.toLowerCase();
```
Comment (coverbeck) [inline src/app/search/facet-search/facet-search-update.pipe.ts:17]: Can searchText be null? This will fail if so.

**Review by coverbeck — APPROVED** [review]

**Review by denis-yuen — APPROVED** [review]

File: src/app/search/facet-search/facet-search-update.pipe.ts (hunk 7 above)
Comment (nrzhao4) [reply src/app/search/facet-search/facet-search-update.pipe.ts:17]: searchText can't be null, if empty it will be an empty string
